frontend/lib/resolution/split-init.cpp

```diff
@@ -227,196 +235,161 @@ void FindSplitInits::handleYield(const uast::Yield* ast, RV& rv) {
-    for (const auto& id : elseFrame->initedVars) {
-      if (elseFrame->eligibleVars.count(id) > 0) {
-        // variable declared in this scope, so save the result
-        allSplitInitedVars.insert(id);
-      } else if (thenFrame->mentionedVars.count(id) == 0) {
-        // variable inited in 'else' and not mentioned in 'then'
-        locInitedVars.insert(id);
-      }
+static void propogateMentions(VarFrame * parentFrame, VarFrame* frame) {
+  for (const auto& id : frame->mentionedVars) {
+    if (frame->declaredVars.count(id) == 0) {
+      parentFrame->mentionedVars.insert(id);
     }
   }
+}
 
-  bool thenReturnsThrows = false;
-  if (thenFrame != nullptr) {
-    thenReturnsThrows = thenFrame->returnsOrThrows;
-  }
-  bool elseReturnsThrows = false;
-  if (elseFrame != nullptr) {
-    elseReturnsThrows = elseFrame->returnsOrThrows;
+void FindSplitInits::errorChecking(const AstNode * node, 
+                                   std::vector<std::vector<ID>> idLists,
+                                   std::vector<std::vector<QualifiedType>> typeLists) {
+  for(size_t i = 1; i < idLists.size(); i++) {
+    CHPL_ASSERT(idLists.at(i).size() == 
+                idLists.at(i-1).size());
+    CHPL_ASSERT(typeLists.at(i).size()  ==
+                typeLists.at(i-i).size());
   }
 
-  std::set<ID> locSplitInitedVars;
-
-  // now consider the initialized 'vars' && propagate them
-  // split init is OK if:
-  //   both then & else blocks initialize it before mentioning it
-  //   one initializes it and the other returns
-  for (const auto& id : locInitedVars) {
-    bool thenInits = false;
-    if (thenFrame) {
-      thenInits = thenFrame->initedVars.count(id) > 0;
-    }
-    bool elseInits = false;
-    if (elseFrame) {
-      elseInits = elseFrame->initedVars.count(id) > 0;
-    }
+  bool orderOk = true;
 
-    if (thenInits && elseInits) {
-      locSplitInitedVars.insert(id);
-    } else if ((thenInits         && elseReturnsThrows) ||
-               (thenReturnsThrows && elseInits)) {
-      // one branch returns or throws and the other inits
-      locSplitInitedVars.insert(id);
-    } else {
-      frame->mentionedVars.insert(id);
+  std::vector<ID> referenceIds = idLists.at(0);
+  for (std::vector<ID> comparedIds : idLists) {
+    for(size_t i = 0; i < referenceIds.size(); i++) {
+      if(referenceIds[i] != comparedIds[i]) {
+        orderOk = false;
+        context->error(node, 
+                       "initialization order does not match between branches");
+        break;
+      }
     }
   }
-
-  // compute set of variables initialized in both branches
-  // (different from locSplitInitedVars because one branch might have
-  //  returned early)
-  // this will be used for error checking
-  std::set<ID> splitInitedInBoth;
-  if (thenFrame && elseFrame) {
-    for (const auto& id : thenFrame->initedVars) {
-      if (elseFrame->initedVars.count(id) > 0) {
-        splitInitedInBoth.insert(id);
+  
+  if (!orderOk) return;
```

Comment:
Seems like you could get rid of this conditional and simply return when you emit the error.